File: project_euler/problem_522/solution.py
"""
Problem 522: Hilbert's Blackout
Grid blackout puzzle on Hilbert curve grids.
"""

def hilbert_d2xy(n: int, d: int):
    """Convert Hilbert curve index d to (x, y) coordinates on 2^n x 2^n grid."""
    x = y = 0
    s = 1
    while s < (1 << n):
        rx = 1 if (d & 2) else 0
        ry = 1 if ((d & 1) ^ rx) else 0
        # Rotate
        if ry == 0:
            if rx == 1:
                x = s - 1 - x
                y = s - 1 - y
            x, y = y, x
        x += s * rx
        y += s * ry
        d >>= 2
        s <<= 1
    return x, y

def hilbert_xy2d(n: int, x: int, y: int):
    """Convert (x, y) coordinates to Hilbert curve index on 2^n x 2^n grid."""
    d = 0
    s = (1 << n) >> 1
    while s > 0:
        rx = 1 if (x & s) > 0 else 0
        ry = 1 if (y & s) > 0 else 0
        d += s * s * ((3 * rx) ^ ry)
        # Rotate
        if ry == 0:
            if rx == 1:
                x = s - 1 - x
                y = s - 1 - y
            x, y = y, x
        s >>= 1
    return d

def generate_hilbert_curve(order: int):
    """Generate all points along a Hilbert curve of given order."""
    n_points = 4 ** order
    points = [hilbert_d2xy(order, d) for d in range(n_points)]
    return points

def blackout_count(order: int, k: int):
    """
    Count lights remaining after blacking out every k-th position
    along the Hilbert curve of given order.
    """
    total = 4 ** order
    blacked = total // k
    return total - blacked

# Compute blackout counts
for order in range(1, 8):
    total = 4 ** order
    for k in [2, 3, 5]:
        remaining = blackout_count(order, k)
        print(f"Order {order} ({total} cells), blackout every {k}: {remaining} remaining")

File: project_euler/problem_522/solution.cpp
#include <bits/stdc++.h>
using namespace std;

// Hilbert curve coordinate conversion

pair<int,int> d2xy(int n, int d) {
    int x = 0, y = 0;
    for (int s = 1; s < (1 << n); s <<= 1) {
        int rx = (d & 2) ? 1 : 0;
        int ry = ((d & 1) ^ rx) ? 1 : 0;
        if (ry == 0) {
            if (rx == 1) {
                x = s - 1 - x;
                y = s - 1 - y;
            }
            swap(x, y);
        }
        x += s * rx;
        y += s * ry;
        d >>= 2;
    }
    return {x, y};
}

int xy2d(int n, int x, int y) {
    int d = 0;
    for (int s = (1 << n) >> 1; s > 0; s >>= 1) {
        int rx = (x & s) > 0 ? 1 : 0;
        int ry = (y & s) > 0 ? 1 : 0;
        d += s * s * ((3 * rx) ^ ry);
        if (ry == 0) {
            if (rx == 1) {
                x = s - 1 - x;
                y = s - 1 - y;
            }
            swap(x, y);
        }
    }
    return d;
}

int main() {
    int order;
    cout << "Enter Hilbert curve order: ";
    cin >> order;

    long long total = 1LL << (2 * order);
    cout << "Total cells: " << total << endl;

    // Example: count lights remaining after every k-th blackout
    for (int k : {2, 3, 5, 7}) {
        long long blacked = total / k;
        long long remaining = total - blacked;
        cout << "Blackout every " << k << ": " << remaining << " remaining" << endl;
    }

    // Verify coordinate mapping for small order
    if (order <= 4) {
        int side = 1 << order;
        cout << "\nHilbert curve path:" << endl;
        for (int d = 0; d < total; d++) {
            auto [x, y] = d2xy(order, d);
            int d2 = xy2d(order, x, y);
            if (d != d2) {
                cout << "ERROR: d=" << d << " -> (" << x << "," << y << ") -> " << d2 << endl;
            }
        }
        cout << "Coordinate mapping verified." << endl;
    }

    return 0;
}
